openmls/src/treesync/mod.rs

```diff
@@ -79,24 +81,77 @@ pub use node::{leaf_node::LeafNode, parent_node::ParentNode, Node};
 pub mod tests_and_kats;
 
 /// An exported ratchet tree as used in, e.g., [`GroupInfo`](crate::messages::group_info::GroupInfo).
-#[derive(
-    PartialEq,
-    Eq,
-    Clone,
-    Debug,
-    Default,
-    Serialize,
-    Deserialize,
-    TlsSerialize,
-    TlsDeserialize,
-    TlsSize,
-)]
-// TODO(#1305): Make the ratchet tree non-malleable.
-//
-// * ValSemXXX: "If the tree has 2^d leaves, then it has 2^(d+1) - 1 nodes. The ratchet_tree vector logically has this number of entries, but the sender SHOULD NOT include blank nodes after the last non-blank node."
-// * ValSemXXX: "Regardless of how the client obtains the tree, the client MUST verify that the root hash of the ratchet tree matches the tree_hash of the GroupContext before using the tree for MLS operations."
+#[derive(PartialEq, Eq, Clone, Debug, Serialize, Deserialize, TlsSerialize, TlsSize)]
 pub struct RatchetTree(Vec<Option<Node>>);
 
+/// An error during processing of an incoming ratchet tree.
+#[derive(Error, Debug, PartialEq, Clone)]
+pub enum RatchetTreeError {
+    /// The ratchet tree is empty.
+    #[error("The ratchet tree has no nodes.")]
+    MissingNodes,
+    /// The ratchet tree has a trailing blank node.
+    #[error("The ratchet tree has trailing blank nodes.")]
+    TrailingBlankNodes,
+}
+
+impl RatchetTree {
+    /// Create a [`RatchetTree`] from a vector of nodes stripping all trailing blank nodes.
+    ///
+    /// Note: The caller must ensure to call this with a vector that is *not* empty after removing all trailing blank nodes.
+    fn from_nodes(mut nodes: Vec<Option<Node>>) -> Self {
```

Comment:
It is a bit unfortunate that we have two ways to create a `RatchetTree` for now. `TryFrom` fails on non-normalized `Vec`s, but this function does the normalization to the extend that empty vectors are accepted.

Reply:
None of the functions accepts trailing blank nodes. The difference lies in the handling of *empty* vectors. `TryFrom` rejects them. `from_nodes` accepts them (because doing otherwise would lead to too much churn).

In any case, i think we do want to have both functions:

* During deserialization we always want to error-out when given trailing blank nodes. Thus, we have `TryFrom`.
* Internally, we want to have a constructor that accepts a `Vec<Option<Node>>`, does the trimming, and returns a valid `RatchetTree`. Thus, we have `RatchetTree::from_nodes`. (That I should probably rename to `from_nodes_trim`.)

Reply:
Discussed off-GitHub.